### DISCUSSION TITLE
Help with writing "no carets in DependencyRange" constraint

### DISCUSSION THREAD

**Discussion by jtoar** [yarnpkg/berry#3570] [Q&A] [answered]
Thoroughly enjoying yarn 3! And trying to enjoy constraints (I’m totally new to prolog but find it pretty cool. It’s like math). I'm almost there (I think), I just needed a little help.

The constraint I want to write is “a dependency shouldn’t have a caret”; a better way to say it may be “dependencies should be static”. Regardless, what I’ve got so far is:

```
yarn constraints query “workspace_has_dependency(WorkspaceCwd, DependencyRange, DependencyType), atom_concat(‘^’, _, DependencyRange)”
```

This lists all the dependencies that have a caret, like…

```
➤ YN0000: ┌ WorkspaceCwd    = 'packages/testing'
➤ YN0000: │ DependencyIdent = '@redwoodjs/auth'
➤ YN0000: │ DependencyRange = '^0.37.4'
➤ YN0000: └ DependencyType  = dependencies
➤ YN0000: Done in 0s 114ms
```

Since I got something to work with `yarn constraints query` I feel like I’m close. But now I don’t know what to put for `gen_enforced_dependency`. I've tried...

```prolog
gen_enforced_dependency(WorkspaceCwd, DependencyIdent, DependencyRange, DependencyType)
  workspace_has_dependency(WorkspaceCwd, DependencyRange, DependencyType), 
  atom_concat(‘^’, _, DependencyRange)
```

But that doesn't seem to be working. 🤔 

Appreciate any help; thanks!

**Comment by merceyz** [comment, marked as answer]
The query you included doesn't seem to match what you ran as you're missing a `DependencyIdent` so you should be getting this error
```
$ yarn constraints query "workspace_has_dependency(WorkspaceCwd, DependencyRange, DependencyType), atom_concat('^', _, DependencyRange)"
➤ YN0044: Existence error: procedure workspace_has_dependency/3 not found (in top_level/0)
➤ YN0000: Failed with errors in 0s 118ms
```

Ignoring that, you need to store the result from `atom_concat` to an atom that `gen_enforced_dependency` can use to set the new range
```prolog
gen_enforced_dependency(WorkspaceCwd, DependencyIdent, TargetDependencyRange, DependencyType) :-
  workspace_has_dependency(WorkspaceCwd, DependencyIdent, CurrentDependencyRange, DependencyType),
  atom_concat('^', TargetDependencyRange, CurrentDependencyRange).
```
https://www.swi-prolog.org/pldoc/man?predicate=atom_concat/3

**Reply by jtoar** [reply]
> The query you included doesn't seem to match what you ran as you're missing a DependencyIdent which gives you this error

Ah you're absolutely right I missed that in my code block.

And that code block you posted works—thanks! So by saving and giving `TargetDependencyRange` to `gen_enforced_dependency`, that tells `yarn constraints` what should actually be the range for that dependency for that workspace . Think I'm getting it; awesome.
